smithy-model/src/main/java/software/amazon/smithy/model/transform/RenameShapes.java

```diff
@@ -0,0 +1,185 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.transform;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.node.StringNode;
+import software.amazon.smithy.model.shapes.AbstractShapeBuilder;
+import software.amazon.smithy.model.shapes.ListShape;
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.SetShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.ShapeVisitor;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.model.traits.DynamicTrait;
+import software.amazon.smithy.model.traits.IdRefTrait;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.utils.Pair;
+
+/**
+ *  Renames shapes using ShapeId pairs while ensuring that the
+ *  transformed model is in a consistent state.
+ *
+ *  This transformer ensures that when an aggregate shape is renamed, all
+ *  members are updated in the model.
+ *
+ * @see ModelTransformer#renameShapes
+ */
+final class RenameShapes {
+    private final Map<ShapeId, ShapeId> renamed;
+
+    RenameShapes(Map<ShapeId, ShapeId> renamed) {
+        this.renamed = renamed;
+    }
+
+    Model transform(ModelTransformer transformer, Model model) {
+        renamed.keySet().removeIf(fromId -> fromId.equals(renamed.get(fromId)));
+        if (renamed.isEmpty()) {
+            return model;
+        }
+        Model.Builder builder = createRenamedModelBuilder(model, renamed);
+
+        getUpdatedIdRefShapes(model, renamed).forEach(builder::addShape);
+
+        return transformer.removeShapes(builder.build(), getShapesToRemove(model, renamed));
+    }
+
+    private Model.Builder createRenamedModelBuilder(Model model, Map<ShapeId, ShapeId> renamed) {
+        Model.Builder builder = model.toBuilder();
+        renamed.forEach((fromId, toId) -> {
+            RenameShapeVisitor renameShapeVisitor = new RenameShapeVisitor(toId);
+            Shape newShape = model.expectShape(fromId).accept(renameShapeVisitor);
+            builder.addShape(newShape);
+            builder.addShapes(newShape.members());
+        });
+        return builder;
+    }
+
+    private Set<Shape> getShapesToRemove(Model model, Map<ShapeId, ShapeId> renamed) {
+        return renamed.keySet().stream()
+                .map(model::expectShape).collect(Collectors.toSet());
+    }
+
+    private Set<Shape> getUpdatedIdRefShapes(Model model, Map<ShapeId, ShapeId> renamed) {
```

Comment:
This code isn't self-documenting (most code isn't). Please add comments that help walk the reader through what's happening and why it's happening.